File: core/flow_executor.py
import json
import os
import time
from typing import Dict, List, Optional, Any, Callable, Tuple
from .flow_diagram import FlowDiagram, FlowNode, NodeType, ConnectionType, Connection
from .actions import Action
from .player import Player, PlayerState
from .exporter import Exporter
from utils.config import Config


class FlowExecutionContext:
    def __init__(self):
        self.variables: Dict[str, Any] = {}
        self.script_results: Dict[str, Any] = {}
        self.current_node_id: Optional[str] = None
        self.execution_path: List[str] = []
        self.loop_counters: Dict[str, int] = {}
    
    def set_variable(self, name: str, value: Any):
        self.variables[name] = value
    
    def get_variable(self, name: str, default: Any = None) -> Any:
        return self.variables.get(name, default)
    
    def set_script_result(self, node_id: str, result: Any):
        self.script_results[node_id] = result
    
    def get_script_result(self, node_id: str) -> Any:
        return self.script_results.get(node_id)


class FlowExecutor:
    def __init__(self, flow: FlowDiagram):
        self._flow = flow
        self._context = FlowExecutionContext()
        self._player: Optional[Player] = None
        self._stop_flag = False
        self._pause_flag = False
        
        self._callbacks: Dict[str, List[Callable]] = {
            'on_node_start': [],
            'on_node_end': [],
            'on_flow_start': [],
            'on_flow_end': [],
            'on_error': [],
            'on_variable_change': [],
        }
        
        for var_name, var in flow.variables.items():
            self._context.set_variable(var_name, var.value)
    
    def add_callback(self, event: str, callback: Callable):
        if event in self._callbacks:
            self._callbacks[event].append(callback)
    
    def _emit(self, event: str, *args, **kwargs):
        for callback in self._callbacks.get(event, []):
            try:
                callback(*args, **kwargs)
            except Exception as e:
                print(f"[FlowExecutor] Callback error: {e}")
    
    def set_player(self, player: Player):
        self._player = player
    
    def stop(self):
        self._stop_flag = True
        if self._player:
            self._player.stop()
    
    def pause(self):
        self._pause_flag = True
        if self._player:
            self._player.toggle_pause()
    
    def resume(self):
        self._pause_flag = False
        if self._player:
            self._player.toggle_pause()
    
    def execute(self) -> Tuple[bool, str]:
        self._stop_flag = False
        self._pause_flag = False
        self._context = FlowExecutionContext()
        
        for var_name, var in self._flow.variables.items():
            self._context.set_variable(var_name, var.value)
        
        self._emit('on_flow_start', self._flow)
        
        valid, errors = self._flow.validate()
        if not valid:
            self._emit('on_error', None, "流程验证失败: " + "; ".join(errors))
            return False, "流程验证失败"
        
        try:
            success = self._execute_from_node("start")
            message = "执行完成" if success else "执行中断"
            self._emit('on_flow_end', success, message)
            return success, message
        except Exception as e:
            self._emit('on_error', None, str(e))
            return False, str(e)
    
    def _execute_from_node(self, node_id: str) -> bool:
        if self._stop_flag:
            return False
        
        while self._pause_flag:
            time.sleep(0.1)
            if self._stop_flag:
                return False
        
        node = self._flow.nodes.get(node_id)
        if not node:
            return False
        
        self._context.current_node_id = node_id
        self._context.execution_path.append(node_id)
        self._emit('on_node_start', node)
        
        success = True
        try:
            if node.node_type == NodeType.START:
                pass
            elif node.node_type == NodeType.END:
                self._emit('on_node_end', node, True)
                return True
            elif node.node_type == NodeType.SCRIPT:
                success = self._execute_script_node(node)
            elif node.node_type == NodeType.CONDITION:
                success = self._execute_condition_node(node)
            elif node.node_type == NodeType.LOOP:
                success = self._execute_loop_node(node)
            elif node.node_type == NodeType.VARIABLE:
                success = self._execute_variable_node(node)
            elif node.node_type == NodeType.DELAY:
                success = self._execute_delay_node(node)
        except Exception as e:
            self._emit('on_error', node, str(e))
            success = False
        
        self._emit('on_node_end', node, success)
        
        if not success:
            return False
        
        next_nodes = self._flow.get_next_nodes(node_id)
        for next_node_id in next_nodes:
            if not self._execute_from_node(next_node_id):
                return False
        
        return True
    
    def _execute_script_node(self, node: FlowNode) -> bool:
        script_path = node.properties.get('script_path', '')
        if not script_path:
            return False
        
        params = node.properties.get('params', {})
        for key, value in params.items():
            if isinstance(value, str) and value.startswith('$'):
                var_name = value[1:]
                params[key] = self._context.get_variable(var_name)
        
        local_group_manager = None
        try:
            from .action_group import LocalActionGroupManager
            local_group_manager = LocalActionGroupManager()
        except:
            pass
        
        result = Exporter.import_from_json(script_path, local_group_manager)
        if result is None:
            return False
        
        actions = result if isinstance(result, list) else result.get('actions', [])
        if not actions:
            return True
        
        for action in actions:
            for key, value in params.items():
                if hasattr(action, 'params') and key in action.params:
                    action.params[key] = value
        
        if self._player is None:
            self._player = Player(tab_key="flow")
        
        self._player.set_actions(actions)
        if local_group_manager:
            self._player.set_local_group_manager(local_group_manager)
        
        self._player.play()
        
        while self._player._state not in [PlayerState.IDLE]:
            if self._stop_flag:
                self._player.stop()
                return False
            time.sleep(0.1)
        
        output_var = node.properties.get('output_variable', '')
        if output_var:
            self._context.set_variable(output_var, True)
            self._emit('on_variable_change', output_var, True)
        
        return True
    
    def _execute_condition_node(self, node: FlowNode) -> bool:
        condition_type = node.properties.get('condition_type', 'variable')
        
        if condition_type == 'variable':
            var_name = node.properties.get('variable_name', '')
            operator = node.properties.get('operator', '==')
            compare_value = node.properties.get('compare_value', '')
            
            var_value = self._context.get_variable(var_name)
            
            if operator == '==':
                result = str(var_value) == str(compare_value)
            elif operator == '!=':
                result = str(var_value) != str(compare_value)
            elif operator == '>':
                try:
                    result = float(var_value) > float(compare_value)
                except:
                    result = False
            elif operator == '<':
                try:
                    result = float(var_value) < float(compare_value)
                except:
                    result = False
            elif operator == '>=':
                try:
                    result = float(var_value) >= float(compare_value)
                except:
                    result = False
            elif operator == '<=':
                try:
                    result = float(var_value) <= float(compare_value)
                except:
                    result = False
            elif operator == 'contains':
                result = str(compare_value) in str(var_value)
            elif operator == 'exists':
                result = var_value is not None
            else:
                result = False
            
            self._context.set_variable('_condition_result', result)
            
        elif condition_type == 'script_result':
            script_node_id = node.properties.get('script_node_id', '')
            result = self._context.get_script_result(script_node_id)
            self._context.set_variable('_condition_result', result is not None and result)
        
        else:
            self._context.set_variable('_condition_result', False)
        
        return True
    
    def _execute_loop_node(self, node: FlowNode) -> bool:
        loop_type = node.properties.get('loop_type', 'count')
        max_iterations = node.properties.get('max_iterations', 10)
        
        loop_id = node.id
        self._context.loop_counters[loop_id] = 0
        
        if loop_type == 'count':
            iterations = 0
            while iterations < max_iterations and not self._stop_flag:
                self._context.loop_counters[loop_id] = iterations + 1
                self._context.set_variable('_loop_index', iterations)
                
                loop_body_start = self._find_loop_body_start(node.id)
                if loop_body_start:
                    if not self._execute_from_node(loop_body_start):
                        return False
                
                iterations += 1
        
        elif loop_type == 'condition':
            while not self._stop_flag:
                condition_var = node.properties.get('condition_variable', '')
                condition_value = node.properties.get('condition_value', True)
                
                current_value = self._context.get_variable(condition_var)
                if current_value != condition_value:
                    break
                
                self._context.loop_counters[loop_id] = self._context.loop_counters.get(loop_id, 0) + 1
                
                loop_body_start = self._find_loop_body_start(node.id)
                if loop_body_start:
                    if not self._execute_from_node(loop_body_start):
                        return False
        
        return True
    
    def _find_loop_body_start(self, loop_node_id: str) -> Optional[str]:
        next_nodes = self._flow.get_next_nodes(loop_node_id)
        return next_nodes[0] if next_nodes else None
    
    def _execute_variable_node(self, node: FlowNode) -> bool:
        var_name = node.properties.get('variable_name', '')
        operation = node.properties.get('operation', 'set')
        value = node.properties.get('value', '')
        
        if operation == 'set':
            if isinstance(value, str) and value.startswith('$'):
                source_var = value[1:]
                value = self._context.get_variable(source_var)
            self._context.set_variable(var_name, value)
        
        elif operation == 'increment':
            current = self._context.get_variable(var_name, 0)
            try:
                self._context.set_variable(var_name, current + 1)
            except:
                pass
        
        elif operation == 'decrement':
            current = self._context.get_variable(var_name, 0)
            try:
                self._context.set_variable(var_name, current - 1)
            except:
                pass
        
        elif operation == 'append':
            current = self._context.get_variable(var_name, [])
            if not isinstance(current, list):
                current = [current]
            current.append(value)
            self._context.set_variable(var_name, current)
        
        self._emit('on_variable_change', var_name, self._context.get_variable(var_name))
        return True
    
    def _execute_delay_node(self, node: FlowNode) -> bool:
        duration = node.properties.get('duration', 1.0)
        start_time = time.time()
        
        while (time.time() - start_time) < duration:
            if self._stop_flag:
                return False
            while self._pause_flag:
                time.sleep(0.1)
                if self._stop_flag:
                    return False
            time.sleep(0.05)
        
        return True
    
    def get_context(self) -> FlowExecutionContext:
        return self._context

File: rust-workspace/crates/simpleRPA-core/src/flow_executor.rs
use crate::flow::{FlowDiagram, FlowNode, NodeType};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

pub struct FlowExecutionContext {
    pub variables: HashMap<String, String>,
    pub script_results: HashMap<String, bool>,
    pub current_node_id: Option<String>,
    pub execution_path: Vec<String>,
    pub loop_counters: HashMap<String, i32>,
}

impl FlowExecutionContext {
    pub fn new() -> Self {
        Self {
            variables: HashMap::new(),
            script_results: HashMap::new(),
            current_node_id: None,
            execution_path: Vec::new(),
            loop_counters: HashMap::new(),
        }
    }

    pub fn set_variable(&mut self, name: &str, value: &str) {
        self.variables.insert(name.to_string(), value.to_string());
    }

    pub fn get_variable(&self, name: &str) -> Option<&str> {
        self.variables.get(name).map(|s| s.as_str())
    }

    pub fn set_script_result(&mut self, node_id: &str, result: bool) {
        self.script_results.insert(node_id.to_string(), result);
    }
}

pub struct FlowExecutor {
    flow: FlowDiagram,
    context: FlowExecutionContext,
    stop_flag: Arc<Mutex<bool>>,
    pause_flag: Arc<Mutex<bool>>,

    on_node_start: Option<Box<dyn Fn(&FlowNode) + Send>>,
    on_node_end: Option<Box<dyn Fn(&FlowNode, bool) + Send>>,
    on_flow_start: Option<Box<dyn Fn(&FlowDiagram) + Send>>,
    on_flow_end: Option<Box<dyn Fn(bool, &str) + Send>>,
    on_error: Option<Box<dyn Fn(Option<&FlowNode>, &str) + Send>>,
}

impl FlowExecutor {
    pub fn new(flow: FlowDiagram) -> Self {
        let mut context = FlowExecutionContext::new();
        for (name, var) in &flow.variables {
            context.set_variable(name, &var.value.to_string());
        }

        Self {
            flow,
            context,
            stop_flag: Arc::new(Mutex::new(false)),
            pause_flag: Arc::new(Mutex::new(false)),
            on_node_start: None,
            on_node_end: None,
            on_flow_start: None,
            on_flow_end: None,
            on_error: None,
        }
    }

    pub fn set_on_node_start<F: Fn(&FlowNode) + Send + 'static>(&mut self, f: F) {
        self.on_node_start = Some(Box::new(f));
    }

    pub fn set_on_node_end<F: Fn(&FlowNode, bool) + Send + 'static>(&mut self, f: F) {
        self.on_node_end = Some(Box::new(f));
    }

    pub fn set_on_flow_start<F: Fn(&FlowDiagram) + Send + 'static>(&mut self, f: F) {
        self.on_flow_start = Some(Box::new(f));
    }

    pub fn set_on_flow_end<F: Fn(bool, &str) + Send + 'static>(&mut self, f: F) {
        self.on_flow_end = Some(Box::new(f));
    }

    pub fn set_on_error<F: Fn(Option<&FlowNode>, &str) + Send + 'static>(&mut self, f: F) {
        self.on_error = Some(Box::new(f));
    }

    pub fn stop(&self) {
        *self.stop_flag.lock().unwrap() = true;
    }

    pub fn pause(&self) {
        *self.pause_flag.lock().unwrap() = true;
    }

    pub fn resume(&self) {
        *self.pause_flag.lock().unwrap() = false;
    }

    pub fn execute(&mut self) -> Result<(), String> {
        *self.stop_flag.lock().unwrap() = false;
        *self.pause_flag.lock().unwrap() = false;
        self.context = FlowExecutionContext::new();

        for (name, var) in &self.flow.variables {
            self.context.set_variable(name, &var.value.to_string());
        }

        if let Some(ref f) = self.on_flow_start {
            f(&self.flow);
        }

        if let Err(errors) = self.flow.validate() {
            let msg = format!("流程验证失败: {}", errors.join("; "));
            if let Some(ref f) = self.on_error {
                f(None, &msg);
            }
            return Err(msg);
        }

        let success = self.execute_from_node("start").unwrap_or(false);
        let message = if success {
            "执行完成".to_string()
        } else {
            "执行中断".to_string()
        };

        if let Some(ref f) = self.on_flow_end {
            f(success, &message);
        }

        if success {
            Ok(())
        } else {
            Err(message)
        }
    }

    fn execute_from_node(&mut self, node_id: &str) -> Result<bool, String> {
        if *self.stop_flag.lock().unwrap() {
            return Ok(false);
        }

        while *self.pause_flag.lock().unwrap() {
            std::thread::sleep(Duration::from_millis(50));
            if *self.stop_flag.lock().unwrap() {
                return Ok(false);
            }
        }

        let node = match self.flow.nodes.get(node_id) {
            Some(n) => n.clone(),
            None => return Ok(false),
        };

        self.context.current_node_id = Some(node_id.to_string());
        self.context.execution_path.push(node_id.to_string());

        if let Some(ref f) = self.on_node_start {
            f(&node);
        }

        let success = match node.node_type {
            NodeType::Start => true,
            NodeType::End => {
                if let Some(ref f) = self.on_node_end {
                    f(&node, true);
                }
                return Ok(true);
            }
            NodeType::Delay => {
                let duration = node
                    .properties
                    .get("duration")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(1.0);
                let end = Instant::now() + Duration::from_secs_f64(duration);
                while Instant::now() < end {
                    if *self.stop_flag.lock().unwrap() {
                        return Ok(false);
                    }
                    std::thread::sleep(Duration::from_millis(50));
                }
                true
            }
            NodeType::Variable => {
                let var_name = node
                    .properties
                    .get("variable_name")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let operation = node
                    .properties
                    .get("operation")
                    .and_then(|v| v.as_str())
                    .unwrap_or("set");
                let value = node
                    .properties
                    .get("value")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");

                match operation {
                    "set" => {
                        self.context.set_variable(var_name, value);
                    }
                    "increment" => {
                        let current = self
                            .context
                            .get_variable(var_name)
                            .and_then(|v| v.parse::<i32>().ok())
                            .unwrap_or(0);
                        self.context
                            .set_variable(var_name, &(current + 1).to_string());
                    }
                    "decrement" => {
                        let current = self
                            .context
                            .get_variable(var_name)
                            .and_then(|v| v.parse::<i32>().ok())
                            .unwrap_or(0);
                        self.context
                            .set_variable(var_name, &(current - 1).to_string());
                    }
                    _ => {}
                }
                true
            }
            NodeType::Condition => {
                let var_name = node
                    .properties
                    .get("variable_name")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let operator = node
                    .properties
                    .get("operator")
                    .and_then(|v| v.as_str())
                    .unwrap_or("==");
                let compare_value = node
                    .properties
                    .get("compare_value")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");

                let var_value = self.context.get_variable(var_name).unwrap_or("");

                let result = match operator {
                    "==" => var_value == compare_value,
                    "!=" => var_value != compare_value,
                    ">" => {
                        var_value.parse::<f64>().unwrap_or(0.0)
                            > compare_value.parse::<f64>().unwrap_or(0.0)
                    }
                    "<" => {
                        var_value.parse::<f64>().unwrap_or(0.0)
                            < compare_value.parse::<f64>().unwrap_or(0.0)
                    }
                    ">=" => {
                        var_value.parse::<f64>().unwrap_or(0.0)
                            >= compare_value.parse::<f64>().unwrap_or(0.0)
                    }
                    "<=" => {
                        var_value.parse::<f64>().unwrap_or(0.0)
                            <= compare_value.parse::<f64>().unwrap_or(0.0)
                    }
                    "contains" => var_value.contains(compare_value),
                    "exists" => !var_value.is_empty(),
                    _ => false,
                };

                self.context
                    .set_variable("_condition_result", &result.to_string());
                true
            }
            NodeType::Loop => {
                let max_iterations = node
                    .properties
                    .get("max_iterations")
                    .and_then(|v| v.as_i64())
                    .unwrap_or(10) as i32;

                let loop_id = node.id.clone();
                self.context.loop_counters.insert(loop_id.clone(), 0);

                let next_nodes = self.flow.get_next_nodes(&node.id);
                let loop_body_start = next_nodes.first().cloned();

                for i in 0..max_iterations {
                    if *self.stop_flag.lock().unwrap() {
                        return Ok(false);
                    }

                    self.context.loop_counters.insert(loop_id.clone(), i + 1);
                    self.context.set_variable("_loop_index", &i.to_string());

                    if let Some(ref body_start) = loop_body_start {
                        if !self.execute_from_node(body_start)? {
                            return Ok(false);
                        }
                    }
                }
                true
            }
            NodeType::Script => {
                let script_path = node
                    .properties
                    .get("script_path")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                let _ = script_path;
                true
            }
        };

        if let Some(ref f) = self.on_node_end {
            f(&node, success);
        }

        if !success {
            return Ok(false);
        }

        let next_nodes = self.flow.get_next_nodes(node_id);
        for next_id in next_nodes {
            if !self.execute_from_node(&next_id)? {
                return Ok(false);
            }
        }

        Ok(true)
    }

    pub fn get_context(&self) -> &FlowExecutionContext {
        &self.context
    }
}
